client-sniffer/src/main/java/org/elasticsearch/client/sniff/Sniffer.java

```diff
@@ -0,0 +1,249 @@
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.elasticsearch.client.sniff;
+
+import org.apache.commons.logging.Log;
+import org.apache.commons.logging.LogFactory;
+import org.apache.http.HttpHost;
+import org.apache.http.config.Registry;
+import org.apache.http.impl.client.CloseableHttpClient;
+import org.elasticsearch.client.RestClient;
+
+import java.io.Closeable;
+import java.io.IOException;
+import java.util.List;
+import java.util.Objects;
+import java.util.concurrent.Executors;
+import java.util.concurrent.ScheduledExecutorService;
+import java.util.concurrent.ScheduledFuture;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicBoolean;
+
+/**
+ * Class responsible for sniffing nodes from an elasticsearch cluster and setting them to a provided instance of {@link RestClient}.
+ * Must be created via {@link Builder}, which allows to set all of the different options or rely on defaults.
+ * A background task fetches the nodes from elasticsearch and updates them periodically.
+ * Supports sniffing on failure, meaning that the client will notify the sniffer at each host failure, so that nodes can be updated
+ * straightaway.
+ */
+public final class Sniffer extends RestClient.FailureListener implements Closeable {
+
+    private static final Log logger = LogFactory.getLog(Sniffer.class);
+
+    private final boolean sniffOnFailure;
+    private final Task task;
+
+    private Sniffer(RestClient restClient, long sniffRequestTimeout, String scheme, long sniffInterval,
+                   boolean sniffOnFailure, long sniffAfterFailureDelay) {
+        HostsSniffer hostsSniffer = new HostsSniffer(restClient, sniffRequestTimeout, scheme);
+        this.task = new Task(hostsSniffer, restClient, sniffInterval, sniffAfterFailureDelay);
+        this.sniffOnFailure = sniffOnFailure;
+        restClient.setFailureListener(this);
+    }
+
+    @Override
+    public void onFailure(HttpHost host) throws IOException {
+        if (sniffOnFailure) {
+            //re-sniff immediately but take out the node that failed
+            task.sniffOnFailure(host);
+        }
+    }
+
+    @Override
+    public void close() throws IOException {
+        task.shutdown();
+    }
+
+    private static class Task implements Runnable {
+        private final HostsSniffer hostsSniffer;
+        private final RestClient restClient;
+
+        private final long sniffInterval;
+        private final long sniffAfterFailureDelay;
+        private final ScheduledExecutorService scheduledExecutorService;
+        private final AtomicBoolean running = new AtomicBoolean(false);
+        private volatile long nextSniffDelay;
+        private volatile ScheduledFuture<?> scheduledFuture;
```

Comment:
just make all the access to `scheduledFuture` synchronized and that way we don't need to make it volatile? it's easyer to undestand I think